### DISCUSSION TITLE
@Subscription on Apollo v2 and v3

### DISCUSSION THREAD

**Discussion by AaronNGray** [MichalLytek/type-graphql#1018] [General]
I have some near standard code hear taken from https://github.com/the-road-to-graphql/fullstack-apollo-subscription-example which is version 2 and I modified to work with version 3.

https://github.com/AaronNGray/fullstack-apollo-3.0-subscription-example

I have modified to use type-graphql. What I am doing wrong it might just be in the @Subscription code on v2 :-
https://github.com/AaronNGray/fullstack-apollo-2.1.0-subscription-type-graphql-example

I have also modified the version 3 code to use type-graphql :-
https://github.com/AaronNGray/fullstack-apollo-3.0-subscription-type-graphql-example/

I am getting the same results on each with the type-graphql subscription code not working. Its probably down to my code on v2 :-
```
import { InputType, ObjectType, Field, Resolver, Query, Subscription, Root, Arg } from 'type-graphql';
import 'reflect-metadata';

@ObjectType()
export class Message {
    @Field(type => Number)
    id: Number = 0
    @Field(type => String)
    content: String = ""
};

@InputType()
export class MessageInput {
    @Field(type => Number)
    id: Number = 0
    @Field(type => String)
    content: String = ""
};

@Resolver(Message)
export class MessageResolver {
    @Query(returns => [Message])
    messages(): Message[] {
        return [
            { id: 0, content: 'Hello!' },
            { id: 1, content: 'Bye!' },
        ];
    }

    @Subscription(returns => Message, {
        topics: "MESSAGE_CREATED"
    })
    messageCreated(
        @Arg("message", type => MessageInput) message : Message
    ): Message {
        console.log(message);
        return message;
    }
};
```

**Comment by MichalLytek** [comment]
> I have also modified the version 3 code to use type-graphql :-

I don't see TypeGraphQL here:
https://github.com/AaronNGray/fullstack-apollo-3.0-subscription-type-graphql-example/blob/main/server/src/index.js

**Comment by AaronNGray** [comment]
@MichalLytek - Sorry made a mistake, fixed now !

**Reply by MichalLytek** [reply]
![image](https://user-images.githubusercontent.com/10618781/132914754-2a6dd7b1-3b50-4b7a-8335-1ee1cb5940bf.png)

Sorry, I'm out. Your code does not compile. Try to prepare better examples, otherwise no one will help you.

**Comment by AaronNGray** [comment]
@MichalLytek - Very sorry its the TypeScript version; should I have noticed and said.

Heres, the sources I have been working with all derived from the original project, and Apollo documentation :-

https://github.com/the-road-to-graphql/fullstack-apollo-subscription-example


https://github.com/AaronNGray/fullstack-apollo-2.1.0-subscription-example (subscriptions working)
https://github.com/AaronNGray/fullstack-apollo-3.0-subscription-example (subscriptions working)

https://github.com/AaronNGray/fullstack-apollo-2.1.0-subscription-type-graphql-example
https://github.com/AaronNGray/fullstack-apollo-2.1.0-subscription-type-graphql-example/blob/main/server/src/index.ts
https://github.com/AaronNGray/fullstack-apollo-2.1.0-subscription-type-graphql-example/blob/main/server/src/types.ts

https://github.com/AaronNGray/fullstack-apollo-3.0-subscription-type-graphql-example
https://github.com/AaronNGray/fullstack-apollo-3.0-subscription-type-graphql-example/blob/main/server/src/index.ts
https://github.com/AaronNGray/fullstack-apollo-3.0-subscription-type-graphql-example/blob/main/server/src/types.ts

**Reply by MichalLytek** [reply]
https://typegraphql.com/docs/subscriptions.html#using-a-custom-pubsub-system

> All we need to do is create an instance of PubSub according to the package instructions and then provide it to the TypeGraphQL buildSchema options

**Comment by AaronNGray** [comment]
@MichalLytek - Thanks I realized how stupid I was at 6am this morning and got it working. 

I have update the above repos.

Just need to do a proper example for graphql-ws now :)
